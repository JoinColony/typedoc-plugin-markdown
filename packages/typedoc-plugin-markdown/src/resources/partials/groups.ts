import {
  DeclarationReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc';
import { MarkdownThemeContext } from '../..';
import { heading, horizontalRule } from '../../utils/elements';

export const groupsPartial = (
  context: MarkdownThemeContext,
  props: ReflectionGroup[],
) => {
  const md: string[] = [];
  props
    ?.filter((group) => !group.allChildrenHaveOwnDocument())
    .forEach((group) => {
      if (group.categories) {
        group.categories
          ?.filter((category) => !category.allChildrenHaveOwnDocument())
          .forEach((category) => {
            md.push(heading(2, `${group.title} ${category.title}`));
            md.push(getMembers(category));
          });
      } else {
        md.push(heading(2, group.title));
        md.push(getMembers(group));
      }
    });
  return md.join('\n\n');

  function getMembers(item: ReflectionGroup | ReflectionCategory) {
    const md: string[] = [];
    item.children.forEach((child, index) => {
      md.push(context.memberContainerPartial(child as DeclarationReflection));
      if (index !== item.children.length - 1) {
        md.push(horizontalRule());
      }
    });
    return md.join('\n\n');
  }
};
