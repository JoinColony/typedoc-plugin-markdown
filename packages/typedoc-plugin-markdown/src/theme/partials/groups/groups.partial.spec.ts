import { ProjectReflection } from 'typedoc';
import { groupsPartial } from './groups.partial';

describe(`Groups Partial:`, () => {
  let project: ProjectReflection;

  beforeAll(async () => {
    project = global.getProject(__dirname);
  });

  test(`should compile groups with categories'`, () => {
    expect(
      groupsPartial(global.getMockContext(), project.groups),
    ).toMatchSnapshot();
  });

  test(`should compile groups without categories'`, () => {
    expect(
      groupsPartial(global.getMockContext(), project.children[0].groups),
    ).toMatchSnapshot();
  });
});
