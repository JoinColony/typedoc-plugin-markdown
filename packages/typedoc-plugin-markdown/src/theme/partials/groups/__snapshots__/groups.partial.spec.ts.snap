// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Groups Partial: should compile groups with categories' 1`] = `
"## Interfaces: CategoryA

{ memberPartial }

---

{ memberPartial }

## Interfaces: Other

{ memberPartial }

## Variables: CategoryA

{ memberPartial }

---

{ memberPartial }

## Variables: CategoryB

{ memberPartial }

## Variables: Other

{ memberPartial }

---

{ memberPartial }

## Functions: CategoryA

{ memberPartial }

---

{ memberPartial }

---

{ memberPartial }

## Functions: CategoryB

{ memberPartial }

## Functions: Other

{ memberPartial }

---

{ memberPartial }"
`;

exports[`Groups Partial: should compile groups without categories' 1`] = `
"## Properties

{ memberPartial }"
`;
